### PR TITLE
[rrd4j] Improve event handling

### LINKED ISSUE #15220: [rrd4j] Persistence warnings
Since recent persistence optimization by @J-N-K I'm seeing a lot of messages like the one below and so do several others on the forum. Running S3515.

I think it happens when I update(+persist) an item multiple times in one second.
To be clear, it's okay that this will only persist one value per second.
But the remaining updates should not result in WARN level output so at least the log level should be lowered to DEBUG.

```
2023-07-10 10:41:07.473 [WARN ] [org.openhab.persistence.rrd4j.internal.RRD4jPersistenceService   ] - RRD4J does not sup                                               port adding past value this=1688978467, last update=1688978468. Discarding Netzeinspeisung - -8.0
```

### PR DESCRIPTION
**Pull request by J-N-K**
Closes #15220

### CODE REVIEW

**Comment by J-N-K** [comment]
@rkoshak Would you like to test that?

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java [hunk 1]
```diff
@@ -191,24 +191,31 @@ public void store(final Item item, @Nullable final String alias) {
 
         long now = System.currentTimeMillis() / 1000;
 
-        scheduler.schedule(() -> internalStore(name, value, now, true), 0, TimeUnit.SECONDS);
+        scheduler.schedule(() -> internalStore(name, value, null), 0, TimeUnit.SECONDS);
```
Comment (lolodomo) [inline bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:194]: Please remove the line just before that set variable `now` that is no more used.

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java [hunk 2]
```diff
@@ -242,32 +249,30 @@ private synchronized void internalStore(String name, double value, long now, boo
...
             logger.debug("Stored '{}' as value '{}' with timestamp {} in rrd4j database", name, storeValue, now);
         } catch (IllegalArgumentException e) {
             String message = e.getMessage();
-            if (message != null && message.contains("at least one second step is required") && retry) {
-                // we try to store the value one second later
-                ScheduledFuture<?> job = scheduledJobs.get(name);
-                if (job != null) {
-                    job.cancel(true);
-                    scheduledJobs.remove(name);
-                }
```
Comment (lolodomo) [inline bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:259]: Why is it no more needed ?

**Review by lolodomo — CHANGES_REQUESTED** [review]
The proposal looks good, I have just 2 comments.

**Comment by J-N-K** [comment]
I have done some further investigations. There is still an issue if a lot of events are created in a very short time (e.g. by a rule):

```
var item = items.getItem("Num");

for (var i=0; i<10; i++) {
  item.postUpdate(i);
}
```

There is a conceptual issue with that in core because `Item` is passed to the persistence service and updates that are faster than the event processing:

```
19:39:58.201 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 9.0 to 0
19:39:58.202 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 0 to 1
19:39:58.202 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=4, Label=Num, Category=) 
19:39:58.202 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=4, Label=Num, Category=) 
19:39:58.203 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 1 to 2
19:39:58.202 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=5, Label=Num, Category=) 
19:39:58.202 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=4, Label=Num, Category=) 
19:39:58.202 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=4, Label=Num, Category=) 
19:39:58.203 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 2 to 3
19:39:58.203 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 3 to 4
19:39:58.204 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=8, Label=Num, Category=) 
19:39:58.204 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 4 to 5
19:39:58.204 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=8, Label=Num, Category=) 
19:39:58.204 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=8, Label=Num, Category=) 
19:39:58.204 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=8, Label=Num, Category=) 
19:39:58.204 [DEBUG] [rd4j.internal.RRD4jPersistenceService] - Num (Type=NumberItem, State=9, Label=Num, Category=) 
19:39:58.204 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 5 to 6
19:39:58.204 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 6 to 7
19:39:58.205 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 7 to 8
19:39:58.205 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'Num' changed from 8 to 9
```

That aside: storing with `now+1` as timestamp will only help for the first "too fast" update, then we'll have the `last>now` issue again (because `last == now+1`). Increasing the timestamp until we find one that is possible feels also wrong, because that may result in very large offsets.

What would be the best case to handle that? "Replacing" the already stored value is not possible with RRD4J.

**Comment by kaikreuzer** [comment]
> What would be the best case to handle that? "Replacing" the already stored value is not possible with RRD4J.

Maybe we could do something like a "delayed write" - i.e. keep the value in memory, wait for the next second and then write the latest then available update to the db (and discard any additional ones one might have received inbetween).

What I don't understand, though, is what is now different to before? We always had the situation that there are many more updates than rrd4j can persist and we had a solution for it in place - we should figure out how that worked and restore it to behave exactly the same way as before (in terms of what states get persisted with which timestamp).

**Comment by J-N-K** [comment]
It worked because the persistence service blocked the event handler until it successfully wrote a datapoint.

**Comment by kaikreuzer** [comment]
> blocked the event handler until it successfully wrote a datapoint.

This still didn't prevent new updates from occuring. So what happened with those? I'd assume that writing to the db didn't take a second, but new updates came in much faster?

**Comment by J-N-K** [comment]
It just scheduled a new `store` with a timestamp one second later until it succeeds. This also messed up the order of the events but it didn't print warnings.

**Comment by kaikreuzer** [comment]
so we essentially persisted all events and made sure there is always a second inbetween? But we persisted them in almost random order...? Ok, if that was the case, maybe we should not try to replicate exactly that behavior...

> we'll miss quickly changing on/off changes of a switch and the item may look unchanged in persistence also it did change in reality.

If something turns on for a fraction of a second only, we can anyhow not correctly persist this - if we extend it to being on for one second, it might be some wrong information already. With the same argument you could say you would want to see that something turned on 10 times within a second and we simply cannot do this. Imho, the most important thing is that the last persisted state is indeed the then valid one.

**Comment by J-N-K** [comment]
Fine. I'll try to implement that.

**Comment by lolodomo** [comment]
> Imho, the most important thing is that the last persisted state is indeed the then valid one.

I agree.

**Comment by rkoshak** [comment]
I've been thinking on this most of the day and I come back and find a proposed solution. 🎉 

I think it's clear that rrd4j is simply unsuitable for rapidly changing data. We might consider adding a warning to that effect to the readme. If we cover the rapidly changing Switch with docs and make sure that the everyMinute batch doesn't cause problems I think that's as good as it gets for rrd4j.

**Comment by kaikreuzer** [comment]
>  I come back and find a proposed solution. 🎉

I hope you can at least agree to that proposal as well!

**Comment by rkoshak** [comment]
> I hope you can at least agree to that proposal as well!

Absolutely! It certainly makes a lot more sense than randomly saving the events like it was doing before or throwing warnings into the logs which for most end users are inexplicable and unclear how they can be handled.

When events are thrown away because a new one came in before it could be saved, it would be good I think to have a DEBUG level log statement explaining that's what's happening. I wouldn't want it to be WARN because it will freak out a bunch of end users over something most of them wouldn't notice in the first place. Couple that with adding the behavior to the docs and I think we are good.

**Comment by J-N-K** [comment]
I think this should do it.

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java [hunk 3]
```diff
@@ -133,9 +134,136 @@ public static RrdDbPool getDatabasePool() {
         return DATABASE_POOL;
     }
 
+    private final ScheduledFuture<?> storeJob;
+
     @Activate
     public RRD4jPersistenceService(final @Reference ItemRegistry itemRegistry) {
         this.itemRegistry = itemRegistry;
+        storeJob = scheduler.scheduleWithFixedDelay(() -> doStore(false), 1, 1, TimeUnit.SECONDS);
```
Comment (kaikreuzer) [inline bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:142]: Wouldn't this line better be placed in the activate() method behind `active = true`? Then there wouldn't be the need to set a 1 second delay either.

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java [hunk 4]
```diff
@@ -190,11 +321,25 @@ public void store(final Item item, @Nullable final String alias) {
         }
 
         long now = System.currentTimeMillis() / 1000;
+        storageMap.computeIfAbsent(now, t -> new ConcurrentHashMap<>()).put(name, value);
```
Comment (kaikreuzer) [inline bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:324]: Would this be the right place to add debug logging as suggested by @rkoshak?

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java [hunk 5]
```diff
@@ -190,11 +321,25 @@ public void store(final Item item, @Nullable final String alias) {
...
+        while (!storageMap.isEmpty()) {
+            long timestamp = storageMap.firstKey();
+            long now = System.currentTimeMillis() / 1000;
+            if (now > timestamp || force) {
+                // no new elements can be added for this timestamp because we are already past that time or the service
+                // requires forced storing
+                Map<String, Double> values = storageMap.pollFirstEntry().getValue();
+                values.forEach((name, value) -> writePointToDatabase(name, value, timestamp));
+            } else {
+                return;
```
Comment (kaikreuzer) [inline bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:337]: Could we run into trouble here once we have forecasts (or some other reason why we have a timestamp in the future) and this while loop runs in circles as the storageMap is not getting empty any time soon?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Thank you - I've left a few small comments.

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java (hunk 3 above)
Comment (kaikreuzer) [reply bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:142]: Yes, that should work as well.

File: bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java (hunk 5 above)
Comment (kaikreuzer) [reply bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:338]: Ah, I somehow didn't consider the `return` here. Yeah, sure.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!

**Pull request merged**